### PR TITLE
Drop unneeded `authz_assessment` calls

### PR DESCRIPTION
**Pull request by reteps**
# Description

This PR removes the unused `selectAssessments` middleware. I determined it was unused with this regex pattern: `\.assessments(?:[^_]|\n)` and excluding `tests,sync,src/lib/chunks.ts`. The results show a single potential usage (and other false positives) in `apps/prairielearn/src/pages/instructorQuestionSettings/instructorQuestionSettings.html.tsx`, but that ends up not coming from `res.locals`.

It also replaces the sproc call in the grader job with a role evaluation. It finally drops the check from the LTI admin page.

# Testing

None; I'm thinking the test suite and manual inspection should be good enough.

### CODE REVIEW

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts [hunk 1]
```diff
@@ -33,13 +33,12 @@ router.get(
...
-    //
-    // The way we implement this check right now with authz_assessment_instance
-    // is overkill, yes, but is easy and robust (we hope).
-    if (gradingJobRow.aai && !gradingJobRow.aai.authorized) {
+    // 'instanceAuthorized' may be null on course pages.
+    const instanceAuthorized = res.locals.authz_data.has_course_instance_permission_view ?? false;
+    const courseAuthorized = res.locals.authz_data.has_course_permission_preview;
+
+    // course_role >= 'Previewer' OR course_instance_role >= 'Student Data Viewer'
+    if (!instanceAuthorized && !courseAuthorized) {
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts]: I don't believe this is a completely faithful rewrite of the checks from the sprocs. Specifically, it seems like this grants course Previewers access to student data.

I think the missing piece is this:

https://github.com/PrairieLearn/PrairieLearn/blob/c34f0ba7e98d6e0112efd6db782bf154c9b6b3c9/apps/prairielearn/src/sprocs/authz_assessment_instance.sql#L110-L116

There's a huge comment above that that explains why it's necessary, I'd encourage you to read through it.

Comment (reteps) [reply apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts]: I see. What I am getting from this is that if the user is not the owner, then they need also need course instance student viewer permissions. We need to check this for both group assessment instances and non-group assessment instances. 

A suggestion would be to return both `assessment_instances` and the list of group users associated with the instance in `instructorGradingJob.sql`. Then, I have enough information to evaluate this extra check for assessment instances in typescript.

Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts]: Yep, that sounds like a good strategy to me!

Comment (reteps) [reply apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts]: I did this in af31821d8cd7e4c11c52a17668e5fe8d0d6f0f52 with whole row returns. I can pull off smaller pieces if wanted.

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts [hunk 2]
```diff
@@ -72,14 +117,33 @@ router.get(
...
+    });
+
+    if (!authorized) {
+      if (needsStudentDataViewerAccess) {
+        throw new error.HttpStatusError(403, 'Access denied (must be a student data viewer)');
+      }
+      throw new error.HttpStatusError(
+        403,
+        'Access denied (must be a course viewer or course instance previewer)',
+      );
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts]: This looks almost (completely?) identical to the above block. Can we put this all into something like an `assertHasAccess(authz_data, gradingJobRow)` function?

Comment (reteps) [reply apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts]: Done in 896860a1cc523eed6e42776d3d37af1ccf9a3f04

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts [hunk 3]
```diff
@@ -7,14 +7,76 @@ import asyncHandler from 'express-async-handler';
...
+function userNeedsStudentDataViewerAccess(user: RawStaffUser, gradingJobRow: GradingJobRow) {
+  // If we don't have an assessment and assessment instance, they don't need student data viewer access
+  if (gradingJobRow.assessment == null || gradingJobRow.assessment_instance == null) {
+    return false;
+  }
+  if (gradingJobRow.assessment.group_work) {
+    // If the user doesn't match any of the group users, they need student data viewer access
+    return gradingJobRow.assessment_instance_group_users?.every(
+      (groupUser) => groupUser.user_id !== user.user_id,
+    );
```
[the remark this answers is not in the document]
Comment (reteps) [reply apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts:78]: Fixed in [39b9348](https://github.com/PrairieLearn/PrairieLearn/pull/13332/commits/39b9348300045173f21fec65e3ba3f70a83ec159).

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts [hunk 4]
```diff
@@ -7,14 +7,79 @@ import asyncHandler from 'express-async-handler';
...
 
+/**
+ * Asserts that the user has access to the grading job.
+ *
+ * Throws an error if the user does not have access to the grading job.
+ */
+function assertHasAccessToGradingJob(authzData: PageAuthzData, gradingJobRow: GradingJobRow) {
+  const courseAuthorized = authzData.has_course_permission_preview;
+
+  // 'courseInstanceAuthorized' may be null on course pages.
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts]: This comment is incorrect, the optional chaining means it'll never be null.

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts [hunk 5]
```diff
@@ -7,14 +7,79 @@ import asyncHandler from 'express-async-handler';
...
+    return courseAuthorized || courseInstanceAuthorized;
+  });
+
+  if (!authorized) {
+    if (needsStudentDataViewerAccess) {
+      throw new error.HttpStatusError(403, 'Access denied (must be a student data viewer)');
+    }
+    throw new error.HttpStatusError(
+      403,
+      'Access denied (must be a course viewer or course instance previewer)',
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts]: This error message doesn't make sense to me:

- It's checking for course preview permissions, not viewer permissions.
- Course instances do not have a previewer permission.

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts [hunk 6]
```diff
@@ -7,14 +7,79 @@ import asyncHandler from 'express-async-handler';
...
+ */
+function userNeedsStudentDataViewerAccess(user: RawStaffUser, gradingJobRow: GradingJobRow) {
+  // If we don't have an assessment and assessment instance, they don't need student data viewer access
+  if (gradingJobRow.assessment == null || gradingJobRow.assessment_instance == null) {
+    return false;
+  }
+  if (gradingJobRow.assessment.group_work) {
+    if (gradingJobRow.assessment_instance_group_users == null) {
+      return true;
+    }
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts]: 👀 when would we expect the list of group users to be null for a group assessment?

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts (hunk 5 above)
Comment (reteps) [reply apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts]: Fixed in ae2aa7c64a020c868d485fd7b041b2351843a8f6.

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts (hunk 6 above)
Comment (reteps) [reply apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts]: We would not, changed to assert in ae2aa7c64a020c868d485fd7b041b2351843a8f6.

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts (hunk 4 above)
Comment (reteps) [reply apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts]: Fixed in ae2aa7c64a020c868d485fd7b041b2351843a8f6.

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts [hunk 7]
```diff
@@ -7,14 +8,80 @@ import asyncHandler from 'express-async-handler';
...
+  // If we don't have an assessment and assessment instance, they don't need student data viewer access
+  if (gradingJobRow.assessment == null || gradingJobRow.assessment_instance == null) {
+    return false;
+  }
+  if (gradingJobRow.assessment.group_work) {
+    assert(gradingJobRow.assessment_instance_group_users !== null);
+    // If the user doesn't match any of the group users, they need student data viewer access
+    return gradingJobRow.assessment_instance_group_users.some(
+      (groupUser) => groupUser.user_id === user.user_id,
+    );
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts:78]: This was changed since my last review, and this does not look correct. This says that the user needs student data viewer access if the user _is_ a member of the group. This is the opposite of what the comment says.

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts [hunk 8]
```diff
@@ -7,14 +8,80 @@ import asyncHandler from 'express-async-handler';
...
+      return courseAuthorized && courseInstanceAuthorized;
+    }
+    return courseAuthorized || courseInstanceAuthorized;
+  });
+
+  if (!authorized) {
+    if (needsStudentDataViewerAccess) {
+      throw new error.HttpStatusError(
+        403,
+        'Access denied (must be a course previewer and student data viewer)',
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts]: It feels like a regression to require course previewer access in this case. Why is student data viewer permissions not sufficient?

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts [hunk 9]
```diff
@@ -7,14 +8,80 @@ import asyncHandler from 'express-async-handler';
...
+      );
+    }
+    throw new error.HttpStatusError(
+      403,
+      'Access denied (must be a course previewer or student data viewer)',
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts:54]: To me, it doesn't seem like student data viewer permissions should be sufficient to access the grading job results for any non-instance-question grading jobs.

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts (hunk 8 above)
Comment (reteps) [reply apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts]: Oops -- I collapsed `authorized := authorized AND (authz_data->>'has_course_instance_permission_view')::boolean;` is equivalent to `((has_course_instance_permission_view OR has_course_permision_preview) AND (has_course_instance_permission_view)` which is equivalent to `has_course_instance_permission_view` not `has_course_instance_permission_view AND has_course_permision_preview`.

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts (hunk 9 above)
Comment (reteps) [reply apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts:54]: This simply mirrors the existing system -- do you think that was already wrong? Or is this somehow a regression?

Comment (reteps) [reply apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts:54]: Specifically this in `apps/prairielearn/src/sprocs/check_assessment_access.sql`:

```
    -- Override if we are course staff
    IF (course_role >= 'Previewer' OR course_instance_role >= 'Student Data Viewer') THEN
        authorized = TRUE;
 ``

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts (hunk 7 above)
Comment (reteps) [reply apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts:78]: Fixed in ca180ca8e09a15a1401b241400bdc6e077cdfee3 and excellent catch, I must have hit a stray tab-complete, I don't remember changing this.

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts (hunk 8 above)
Comment (reteps) [reply apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts]: Fixed in ca180ca8e09a15a1401b241400bdc6e077cdfee3.

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts (hunk 9 above)
Comment (nwalters512) [reply apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts:54]: Ah, good point that this is existing behavior. I don't think it's correct, but we can fix that elsewhere. I'll stack a PR on top of this one to explore a change.

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.sql [hunk 10]
```diff
@@ -1,23 +1,21 @@
...
-    ELSE to_jsonb(
-      authz_assessment_instance (
-        ai.id,
-        $authz_data,
-        $req_date,
-        ci.display_timezone,
-        a.group_work
-      )
-    )
-  END AS aai,
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.sql:14]: Testing reveals that you need to update the query params that are passed to this.

Image

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts [hunk 11]
```diff
@@ -7,14 +8,77 @@ import asyncHandler from 'express-async-handler';
...
+ * Assessment instances have an additional restriction: they must be accessed by a student data viewer,
+ * or the owner of the assessment instance.
+ */
+function userNeedsStudentDataViewerAccess(user: RawStaffUser, gradingJobRow: GradingJobRow) {
+  // If we don't have an assessment and assessment instance, they don't need student data viewer access
+  if (gradingJobRow.assessment == null || gradingJobRow.assessment_instance == null) {
+    return false;
+  }
+  if (gradingJobRow.assessment.group_work) {
+    assert(gradingJobRow.assessment_instance_group_users !== null);
```
Comment (nwalters512) [inline apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts]: I know a previous comment of mine lead to this assertion. Thinking about this more deeply (and confirmed with manual testing), a group with zero members will result in a `null` array here. This can occur if e.g. all members manually leave the group.

Comment (reteps) [reply apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts]: Oops, I didn't even consider that you can leave a group.

**Comment by reteps** [comment]
CleanShot 2025-11-12 at 15 25 29@2x
Tested this. Also, caf1868 handles the null user case again.

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts (hunk 11 above)
Comment (reteps) [reply apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.ts]: Fixed in [caf1868](https://github.com/PrairieLearn/PrairieLearn/pull/13332/commits/caf18687e847100852eb9ab7cbcb498efd432c5f).

File: apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.sql (hunk 10 above)
Comment (reteps) [reply apps/prairielearn/src/pages/instructorGradingJob/instructorGradingJob.sql:14]: Fixed in [caf1868](https://github.com/PrairieLearn/PrairieLearn/pull/13332/commits/caf18687e847100852eb9ab7cbcb498efd432c5f).

**Review by nwalters512 — APPROVED** [review]
